torch/distributed/fsdp/optim_utils.py

```diff
@@ -192,11 +221,95 @@ def _unflatten_communicated_optim_state(
     return unflat_param_state
 
 
+def _flatten_full_optim_state_dict(
```

Comment:
Common logic to both `shard_full_optim_state_dict()` and `scatter_full_optim_state_dict()`, only differentiated by `shard_state: bool`.